### ISSUE TITLE
[postgrest]: lost somewhat type safety since rework

### ISSUE THREAD

**Issue by Vinzent03** [supabase/supabase-flutter#162]
Coming from supabase-community/supabase-flutter#186 I experienced that you can't specify the result type by generics.
Since the rework, the return type is `dynamic`, because by default it returns the raw postgrest output. But by specifying `count` or `forceResponse` in `FetchOptions` it returns `PostgrestResponse`. I think this makes the developer experience much harder. Although the 1.0.0 release is focused on developer experience.
My proposal is to put `forceResponse` and `count` into its own method in `PostgrestFilterBuilder`. Something like `options(FetchOptions())`. This method would change the generic to be `PostgrestResponse`.
This would be again a breaking change, so I don't know if it's worth it. `FetchOptions` would be removed from `select`,`upsert`,.. `FetchOptions.head` could stay in these methods.

Another case of type safety is whether `PostgrestResponse.data` is a list (default) or a single object (by calling `single()`).

These are just my ideas and thoughts to make this library more type safe. I'm curious what you think. If you don't really know what I mean, I can make a draft pull request.

**Comment by bdlukaa** [comment]
Before, `PostgrestResponse.data` was also dynamic. How would it be different from now?

**Comment by Vinzent03** [comment]
Making `PostgrestResponse.data` more type safe would be an improvement. Now, you don't even know directly if it's `PostgrestResponse` or `PostgrestResponse.data`. I think an extra level is added, which can be removed by chaining methods differently.

**Comment by Vinzent03** [comment]
In addition to breaking change, it may not be a good idea, because future changes might make it impossible to make it type safe.

**Comment by bdlukaa** [comment]
I still don't get how this would be any better than the current implementation, since`PostgrestResponse.data` would be `dynamic` as well. 

We tried to mirror the js implementation by returning the data automatically, and it's a lot more dev-friendly.

**Comment by Vinzent03** [comment]
I see that it is too hard implementing more type safety, and it's not really worth.

**Comment by caseycrogers** [comment]
> My proposal is to put forceResponse and count into its own method in PostgrestFilterBuilder. 

I strongly endorse this. Methods that can return different types depending on the arguments passed into them are very non-idiomatic in most typed languages. The current generics system is also extremely kludgy, especially given that the strong dart analysis options make type parameters that can be statically inferred mandatory.

Using union types or records seems like papering over the issue with non-idiomatic practices. Trying to mirror the JS SDK too closely is going to produce a messy API because a lot of standard practices in JS (eg methods that return union types) are not idiomatic in Dart.

The `own method` proposal is the best because it:
1. Follows Dart best practices
2. Follows the builder pattern that Supabase is already using
3. Provides type safety by default
4. Makes developer errors (ie assuming the output is one type when it's actually another) static analysis errors as opposed to runtime errors

**Comment by caseycrogers** [comment]
One thing that has occurred to me, the above may not be super coherent given that the current query filter builder methods are destructive. As is, they mutate the existing object and then return it. Mutating itself is already kind of an anti pattern anyway, so I suggest everything switch to an immutable pattern instead where copies of the object are returned. This has all kinds of usability advantages anyway.
Here's an example of why the destructive pattern is undesirable. If I want to run a query twice, one with filter A applied, and the other with filter B applied, I can do the following:
```dart
var fooQuery = myQuery.eq('some_col', 'foo');
var notFooQuery = myQuery.neq('some_col', 'foo');
```
This may seem a little contrived (why not just use an `and`?) but it's already come up for me in my project where I wanted to order my query results by whether or not a column equaled a scalar value. That can't be done with `order` because it only accepts values, not expressions, hence the construction above. However, because the query builder mutates itself and there's no `copy` method, I had to make two versions of `myQuery` from scratch.
Not to mention, `select` already follows the `return a new object` pattern so the current behavior is a weird mishmash of destructive functions and non-destructiev functions. The fact that the destructive functions even return `this` is highly confusing because it implies that a new object was created and the old object was left alone. Dart already has the `..` pattern anyway so there's no point in returning `this`. It only serves to confuse the developer by falsely implying that the method is non-destructive.

TLDR: The filter builder methods should either return `void` instead of `this` or be made non-destructive, highly preferably the latter. My proposal above only really makes sense if they're made to be non-destructive because the type parameter of a class can't change after construction.

**Comment by dshukertjr** [comment]
@caseycrogers Great point. I wonder if we should open an issue just to keep track of all the things we want to implement in v2.

**Comment by dshukertjr** [comment]
I will close this one in favor of v2 feature tracker. Let's add any suggestions there!
